plugins/urls/__init__.py

```diff
@@ -335,343 +355,151 @@ def _respond_shorten(self, result, source, handler):
-            This occurs in each of the following cases..
-
-            * When a portscan is detected and stopped
-            * When the page simply has no title
-            * When there is an exception in the chain somewhere
-
-        :param url: The URL to check
-        :param use_handler: Whether to use specialized handlers
+                self.logger.exception("Error fetching short URL.")
+                caller.respond("Error: %s" % e)
 
-        :type url: str
-        :type use_handler: bool
+    @inlineCallbacks
+    def run_handlers(self, _url, context):
+        for priority in reversed(sorted(self.handlers.iterkeys())):
+            for handler in self.handlers[priority]:
+                try:
+                    d = handler.match(_url, context)
+                except Exception as e:
+                    self.logger.exception(
+                        "Error caught while attempting to match "
+                        "handler {0}".format(
+                            handler.name
+                        )
+                    )
 
-        :returns: A tuple containing the result
-        :rtype: tuple(None, None), tuple(str, str), tuple(str, None)
-        """
+                    continue
 
-        domain = ""
-        self.logger.trace(_("Url: %s") % url)
-        try:
-            parsed = urlparse.urlparse(url)
-            domain = parsed.hostname
-
-            ip = socket.gethostbyname(domain)
-
-            matches = all_matching_cidrs(ip, ["10.0.0.0/8", "0.0.0.0/8",
-                                              "172.16.0.0/12",
-                                              "192.168.0.0/16", "127.0.0.0/8"])
-
-            if matches:
-                self.logger.warn(_("Prevented a portscan: %s") % url)
-                return None, None
-
-            if domain.startswith("www."):
-                domain = domain[4:]
-
-            if use_handler:
-                for pattern in self.handlers:
-                    if fnmatch.fnmatch(domain, pattern):
-                        try:
-                            result = self.handlers[domain](url)
-                            if result:
-                                return to_unicode(result), None
-                        except Exception:
-                            self.logger.exception(_("Error running handler, "
-                                                    "parsing title normally."))
-
-            self.logger.trace(_("Parsed domain: %s") % domain)
-
-            request = urllib2.Request(url)
-            if domain in self.spoofing:
-                self.logger.debug(_("Custom spoofing for this domain found."))
-                user_agent = self.spoofing[domain]
-                if user_agent:
-                    self.logger.debug(_("Spoofing user-agent: %s")
-                                      % user_agent)
-                    request.add_header("User-agent", user_agent)
+                if isinstance(d, Deferred):
+                    r = yield d
```

Comment:
Handling the Failures now, but I'd like to add Failure support to .exception() in the logger sometime, or add a .failure()